woocommerce/woocommerce-gutenberg-products-block
# Product Query: pass any product taxonomies existing in the URL parameters.

While working on 12894-gh-Automattic/woocommerce.com I have noticed that there isn't a way to filter products based on custom product taxonomies. The code only allows a static list of `product_cat` and `product_tag`. This PR aims at using any product taxonomy present in the URL parameters to build the `$query_args` for the `WP_Query`.

Related convo: p1648465508181829-slack-C8X6Q7XQU

#### Automated Tests
### Manual Testing

How to test the changes in this Pull Request:

- On your local environment register a new taxonomy
```
$labels = array(
	'name'                       => 'Custom Category',
	'singular_name'              => 'Category',
	'menu_name'                  => 'Categories',
	'all_items'                  => 'All Categories',
	'parent_item'                => 'Parent Category',
	'parent_item_colon'          => 'Parent Category:',
	'new_item_name'              => 'New Category Name',
	'add_new_item'               => 'Add New Category',
	'edit_item'                  => 'Edit Category',
	'update_item'                => 'Update Category',
	'separate_items_with_commas' => 'Separate Category with commas',
	'search_items'               => 'Search Categories',
	'add_or_remove_items'        => 'Add or remove Categories',
	'choose_from_most_used'      => 'Choose from the most used Categories',
);
$taxonomy_args = array(
	'labels'       => $labels,
	'description'  => 'A new custom product taxonomy',
	'hierarchical' => true,
	'public'       => true,
	'show_ui'      => true,
);
register_taxonomy( 'custom_categories', 'product', $taxonomy_args );
```
- Edit a product and add to it a new term of the `custom_categories` taxonomy
- Go to `/wp-json/wc/store/products?_unstable_tax_custom_categories=[that-category-id]` and make sure you get only products where this term was added
- Go to `/wp-json/wc/store/products` and make sure that all products get returned as expected

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).

1. Add an All Products Block to a page, add the Filter Products by Attribute Block, and Filter by Price Block too.
2. Go to the page and use the filters, ensure they update the All Products Block correctly.

### Changelog

Enhancements
> Allow Store API to filter products by custom taxonomies.

Bug Fixes
> Fix `not_in` operator for category, category, and attribute params.

### Release Note
**Store API**
As part of enhancing the Store API, we added a way to filter products by custom defined product taxonomies. The are two new parameters that can be passed to the API and should follow the dynamic patters `_unstable_tax_[product-taxonomy]` and `_unstable_tax_[product-taxonomy]_operator`. They work the same way as `category` and `category_operator` but in this case you should replace `[product-taxonomy]` with any custom defined product taxonomy. eg
`?_unstable_tax_hot_categories=1,2,3&_unstable_tax_hot_categories_operator=in` 
will return all products appearing in `hot_categories` taxonomy for terms 1 or 2 or 3.
